### ISSUE TITLE
BC SaaS issues installing newer versions of AL extensions

### ISSUE THREAD

**Issue by CanuckNav74** [Microsoft/AL#4221]
I have an AL extension and I want to bump up the version# of the .app (in this case, 1.0.0.0 to 1.1.0.0)

My goal is to deploy the extention to my customer's tenant sandbox (which worked just fine for my initial 1.0.0.0 build of my .app). The launch.json launch configuration's schemaUpdateMode=Synchronize

In the extension I also added an extension upgrade codeunit (where subtype = upgrade) and added all the necessary triggers for it (I don't have any code in those triggers yet as I don't have any relevant data to upgrade - my .app only has two .al report objects). 

When deploying the 1.1.0.0 extsion to the tenant sandbox via F5, I get the following error dialog:

```
[Window Title]
Visual Studio Code

[Content]
Could not publish the package to the server. See Visual Studio Code debug console for more details.

[Open launch.json] [Cancel]
```

The debug console however is devoid of any useful details.

So I then try again but this time launch the VS code debugger via F6 (which does give me useful feedback on the error in the VS Code output window) and it reads:

```
Microsoft (R) AL Compiler version 2.0.48254
Copyright (C) Microsoft Corporation. All rights reserved

Compilation started for project 'Atlantica Tenant Customizations' containing '3' files at '10:25:6.246'.


Compilation ended at '10:25:6.579'.

Success: The package is created.
[2018-11-16 10:25:06.58] Sending request to http://microsoft-bc-ca:7049/nav/dev/apps?SchemaUpdateMode=synchronize
[2018-11-16 10:25:10.78] The request for path /nav/dev/apps?SchemaUpdateMode=synchronize failed with code 422. Reason: Cannot install the extension {extension name} by {publisher name} 1.1.0.0 for the tenant default because an earlier version was already installed. Run Start-NAVAppDataUpgrade to upgrade the extension.
```

This error is very misleading as the NAVAppDataUpgrade cmdlet is for docker and onprem only according to my understanding. 

The error is a red herring. When I go to the tenant's system I can see the latest extension of my app via the extension management page (1.1.0.0 in this case) but it's not installed. It's published but not installed. 

So I install it. Subsequent debug sessions from visual studio code work fine at this point. 

Does installing a published extension via the Extension Management page on BC SaaS call the upgrade code within that extension? The documentation for upgrading v2 extensions (here - https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-upgrading-extensions ) does not have any relevant information in the context of BC SaaS. The documentation that is there is for on-prem. 

AL Language extension is 2.0.48254
Tenant is on CA Dynamics NAV 13.0 (25488)

**Comment by CanuckNav74** [comment]
I spoke too soon. looks like the problem is more serious than I feared and I have no suitable workaround.

I reset my app back to 1.0.0.0 and deployed it via schemamode=recreate. It publishes and installs. 

I then up the version in my app.json to 1.1.0.0 and deploy it via schemamode=synchronize. I get the error in the VS Code output window:

Cannot install the extension {app name} by {publisher name} 1.1.0.0 for the tenant msca1a7355t38946840 because an earlier version was already installed. Run Start-NAVAppDataUpgrade to upgrade the extension.

I can see the extension is published (1.1.0.0) but the "install" option is completely grayed out for the extension. 

Bottom line is I cannot do anything with a published (but not installed) extension where I have no option to upgrade it. 

What am I missing??

deploying it via schemamode=recreate every time is not a suitable solution.

**Comment by CanuckNav74** [comment]
funny enough, the "install" option was grayed out on the triple-dot menu for the extension (e.g. the menu that gives options for install, uninstall, unpublish, configure, etc). If I clicked "view" however, I had the option to install it. 

Why is the install option accessible from one area but not from the other?

**Comment by StanislawStempin** [comment]
Running upgrade in sandboxes is not yet supported. You would need to test upgrade in Docker images and using the cmdlets described in the help topics that you mentioned.
On sandboxes you can only deploy the same version of the extension.

**Comment by CanuckNav74** [comment]
@StanislawStempin is there an ETA on when upgrades from the sandbox will be supported? 

Testing via docker is unacceptable for larger tenant-specific extensions where we need actual tenant data to test against. Rapidstart excel uploads is not an option - I was really excited about the sandbox ability to replicate form the production system, but if the sandbox cannot handle .app upgrades then this is a problem.  

Its quite scary to to think that there is no way to safely and quickly test these extension with actual tenant data if the sandbox doesn't offer extension upgrade support.

**Comment by pborring** [comment]
See Idea on this https://experience.dynamics.com/ideas/idea/?ideaid=3cdfaed9-a5ed-e811-a140-0003ff68cd46
